### PR TITLE
Added distributed lock api support to javascript sdk

### PR DESCRIPTION
**Pull request by amulyavarote**
Signed-off-by: Amulya Varote <amulyavarote@microsoft.com>

# Description

_Please explain the changes you've made_

Added distributed lock api support to javascript sdk

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #284 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation

### CODE REVIEW

File: examples/configuration/package.json [hunk 1]
```diff
@@ -21,4 +21,4 @@
         "@dapr/dapr": "file:../../build",
         "@types/node": "^15.3.0"
     }
-}
\ No newline at end of file
+}
```
Comment (shubham1172) [inline examples/configuration/package.json:24]: Nit: revert unrelated changes.

File: examples/distributedLock/README.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
+# Example - Distributed Lock APIs
```
Comment (shubham1172) [inline examples/distributedLock/README.md:1]: Given that in both examples we have both `tryLock` and `unlock` functions being invoked, is it possible to merge the two examples into one by taking a specific scenario and demonstrate both?

File: examples/distributedLock/TryLockApplication/package.json [hunk 3]
```diff
@@ -0,0 +1,24 @@
+{
+    "name": "dapr-example-config",
+    "version": "1.0.0",
+    "description": "An example utilizing the Dapr JS SDK to invoke a service",
+    "main": "dist/index.js",
+    "private": true,
+    "scripts": {
+        "build": "rimraf ./dist && tsc",
+        "start": "npm run build && node dist/index.js",
+        "start:dapr-grpc": "dapr run  --app-id distributedLock --app-protocol grpc --components-path ./Components npm run start",
```
Comment (shubham1172) [inline examples/distributedLock/TryLockApplication/package.json:10]:
```suggestion
        "start:dapr-grpc": "dapr run  --app-id distributedLock --app-protocol grpc --components-path ./components npm run start",
```

File: examples/distributedLock/TryLockApplication/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    process.env.DAPR_GRPC_PORT ?? daprPortDefault,
+    CommunicationProtocolEnum.GRPC
+  );
+
+  const storeName = "redislock";
+  let resourceId = "resourceId";
+  let lockOwner = "owner1";
+  let expiryInSeconds = 1000;
+
+  const tryLockResponse = await client.lock.tryLock(storeName, resourceId, lockOwner, expiryInSeconds);
```
Comment (shubham1172) [inline examples/distributedLock/TryLockApplication/src/index.ts:32]: Maybe some more console.logs so that users know what's happening in the example from the output directly?
E.g., `invoking tryLock on ${storeName}, ${resourceID} as owner: ${lockOwner}`

File: examples/distributedLock/TryLockApplication/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+async function start() {
+  const client = new DaprClient(
+    daprHost,
+    process.env.DAPR_GRPC_PORT ?? daprPortDefault,
+    CommunicationProtocolEnum.GRPC
+  );
+
+  const storeName = "redislock";
+  let resourceId = "resourceId";
+  let lockOwner = "owner1";
```
Comment (shubham1172) [inline examples/distributedLock/TryLockApplication/src/index.ts:29]: const since it's not being changed anywhere

File: examples/distributedLock/TryLockApplication/src/index.ts (hunk 4 above)
Comment (shubham1172) [reply examples/distributedLock/TryLockApplication/src/index.ts:32]: Applicable for all examples.

File: examples/distributedLock/TryLockApplication/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  console.log("Unlock API response when lock is not acquired: " + getResponseStatus(lockUnexistResponse.status));
+
+  resourceId = "resourceId";
+  lockOwner = "owner1";
+  expiryInSeconds = 25;
+  const tryLockResponse1 = await client.lock.tryLock(storeName, resourceId, lockOwner, expiryInSeconds);
+  console.log("Acquired Lock? " + tryLockResponse1.success);
+
+  await new Promise(resolve => setTimeout(resolve, 20000));
+
```
Comment (shubham1172) [inline examples/distributedLock/TryLockApplication/src/index.ts:49]:
```suggestion
```

File: src/interfaces/Client/IClientLock.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { TryLockResponse } from "../../types/lock/TryLockResponse";
+import { UnlockResponse } from "../../types/lock/UnLockResponse";
+
+export default interface IClientLock {
+  tryLock(storeName: string, resourceId: string, lockOwner: string, expiryInSeconds: number): Promise<TryLockResponse>;
+  unLock(storeName: string, resourceId: string, lockOwner: string): Promise<UnlockResponse>;
```
Comment (shubham1172) [inline src/interfaces/Client/IClientLock.ts:19]:
```suggestion
  unlock(storeName: string, resourceId: string, lockOwner: string): Promise<UnlockResponse>;
```

File: src/interfaces/Client/IClientLock.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { TryLockResponse } from "../../types/lock/TryLockResponse";
+import { UnlockResponse } from "../../types/lock/UnLockResponse";
+
+export default interface IClientLock {
```
Comment (shubham1172) [inline src/interfaces/Client/IClientLock.ts:17]: Please add documentation for functions and parameters.

File: src/implementation/Client/HTTPClient/lock.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+import IClientLock from '../../../interfaces/Client/IClientLock';
+
+export default class HTTPClientLock implements IClientLock {
+    client: HTTPClient;
+
+    constructor(client: HTTPClient) {
+        this.client = client;
+    }
+    tryLock(_storeName: string, _resourceId: string, _lockOwner: string, _expiryInSeconds: number): Promise<TryLockResponseResult> {
+        throw new Error('HTTP is currently not supported.');
```
Comment (shubham1172) [inline src/implementation/Client/HTTPClient/lock.ts:26]: This should be done as a separate PR, but it might be cleaner to introduce two new errors - `HttpNotSupportedError` and `GrpcNotSupportedError` and use them in our codebase. Created: https://github.com/dapr/js-sdk/issues/297

File: scripts/fetch-proto.sh [hunk 10]
```diff
@@ -1,7 +1,7 @@
 #!/bin/bash
 OS=$(echo `uname`|tr '[:upper:]' '[:lower:]')
 ARCH=$(uname -m)
-BRANCH_NAME="v1.7.0"
+BRANCH_NAME="master"
```
Comment (shubham1172) [inline scripts/fetch-proto.sh:4]: TODO: update this after v1.8.0 release and refresh proto.

File: package.json [hunk 11]
```diff
@@ -26,6 +26,7 @@
     "author": "Xavier Geerinck",
     "license": "ISC",
     "dependencies": {
+        "@dapr/dapr": "^2.2.1",
```
Comment (shubham1172) [inline package.json:29]: Something wrong with this as it's introducing a dependency on itself!

File: examples/invocation/package.json [hunk 12]
```diff
@@ -21,4 +21,4 @@
         "@dapr/dapr": "file:../../build",
         "@types/node": "^15.3.0"
     }
-}
\ No newline at end of file
+}
```
Comment (shubham1172) [inline examples/invocation/package.json:24]: nit: please revert unrelated changes

File: test/e2e/grpc/client.test.ts [hunk 13]
```diff
@@ -318,4 +319,26 @@ describe('grpc/client', () => {
       await stream2.stop();
     });
   });
-});
+
+  describe('distributed lock', () => {
```
Comment (shubham1172) [inline test/e2e/grpc/client.test.ts:323]: Some more scenarios:
1. two owners are acquiring the same lock at the same time 
2. two owners are acquiring different locks on unique resourceIDs 
3. unlocking a lock that belongs to someone else

**Review by shubham1172 — CHANGES_REQUESTED** [review]
Thank you for the PR @amulyavarote!! Added a few comments.

**Comment by shubham1172** [comment]
Can you also check the e2e tests please? They seem to be failing.

File: test/e2e/grpc/client.test.ts [hunk 14]
```diff
@@ -318,4 +319,45 @@ describe('grpc/client', () => {
...
+
+    it('should be able to acquire a lock when a different lock is acquired by the other owner', async () => {
+      const tryLockOne = await client.lock.tryLock("redislock", "resourceId2", "owner1", 5);
+      expect(tryLockOne.success).toEqual(true);
+      const tryLockTwo = await client.lock.tryLock("redislock", "resourceId3", "owner2", 5);
+      expect(tryLockTwo.success).toEqual(true);
+    });
+
+    it('should not be able to unlock a lock when that lock is acquired by the other owner/process', async () => {
+      const tryLockOne = await client.lock.tryLock("redislock", "resourceId3", "owner2", 5);
```
Comment (shubham1172) [inline test/e2e/grpc/client.test.ts:359]: Is this test incomplete?

File: src/interfaces/Client/IClientLock.ts (hunk 8 above)
Comment (shubham1172) [reply src/interfaces/Client/IClientLock.ts:17]: This is the documentation that SDK users will see when they try to use these functions, so it will be good to have a documentation that tells the user about the different parameters and what to expect, are there any possible exceptions that can be thrown, etc. Example https://github.com/dapr/js-sdk/blob/master/src/types/logger/LoggerOptions.ts

File: src/implementation/Client/GRPCClient/lock.ts [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+            const wrapped: UnLockResponseResult = {
+                status: this.getUnLockResponse(res)
+            }
+    
+            return resolve(wrapped);
+          });
+        });
+      }
+
+    getUnLockResponse(res: UnlockResponse) {
```
Comment (shubham1172) [inline src/implementation/Client/GRPCClient/lock.ts:73]: nit: getUnlockResponse

File: src/types/lock/UnLockResponse.ts [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+*/
+
+export enum LockStatus {
+    Success,
+    LockUnexist,
+    LockBelongToOthers,
+    InternalError
+}
+
+export type UnlockResponse = {
```
Comment (shubham1172) [inline src/types/lock/UnLockResponse.ts:21]: nit: Filename casing

File: examples/distributedLock/UnLockApplication/src/index.ts [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+
+  //Trying to unlock the lock after the lock used by this process(UnLockApplication) is expired
+  unLockResponse = await client.lock.unlock(storeName, resourceId, lockOwner);
+  console.log(`Unlocking on ${storeName}, ${resourceId} as owner: ${lockOwner}`);
+  console.log("Unlock API response when lock is released after the expiry time and lock does not exist: " + getResponseStatus(unLockResponse.status));
+}
+
+function getResponseStatus(status: LockStatus) {
+  switch(status) {
+    case 0:
```
Comment (shubham1172) [inline examples/distributedLock/UnLockApplication/src/index.ts:63]: We should use LockStatus enum instead of numbers.

File: examples/distributedLock/TryLockApplication/src/index.ts [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+  const tryLockResponse1 = await client.lock.tryLock(storeName, resourceId, lockOwner, expiryInSeconds);
+  console.log(`Acquiring lock on ${storeName}, ${resourceId} as owner: ${lockOwner}`);
+  console.log("Acquired Lock? " + tryLockResponse1.success);
+
+  await new Promise(resolve => setTimeout(resolve, 20000));
+}
+
+function getResponseStatus(status: LockStatus) {
+  switch(status) {
+    case 0:
```
Comment (shubham1172) [inline examples/distributedLock/TryLockApplication/src/index.ts:57]: Same comment here for using enum instead of number.

File: examples/distributedLock/TryLockApplication/src/index.ts (hunk 5 above)
Comment (shubham1172) [reply examples/distributedLock/TryLockApplication/src/index.ts:29]: This still requires your attention

File: examples/distributedLock/TryLockApplication/src/index.ts [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+  console.log("Unlock API response when lock is not acquired: " + getResponseStatus(lockUnexistResponse.status));
+
+  resourceId = "resourceId";
+  lockOwner = "owner1";
+  expiryInSeconds = 25;
+  const tryLockResponse1 = await client.lock.tryLock(storeName, resourceId, lockOwner, expiryInSeconds);
+  console.log(`Acquiring lock on ${storeName}, ${resourceId} as owner: ${lockOwner}`);
+  console.log("Acquired Lock? " + tryLockResponse1.success);
+
+  await new Promise(resolve => setTimeout(resolve, 20000));
```
Comment (shubham1172) [inline examples/distributedLock/TryLockApplication/src/index.ts:52]: What is this for?

File: examples/distributedLock/TryLockApplication/src/index.ts [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+  //Checking if the lock exists.
+  const lockUnexistResponse = await client.lock.unlock(storeName, resourceId, lockOwner);
+  console.log(`Unlocking on ${storeName}, ${resourceId} as owner: ${lockOwner}`);
+  console.log("Unlock API response when lock is not acquired: " + getResponseStatus(lockUnexistResponse.status));
+
+  resourceId = "resourceId";
+  lockOwner = "owner1";
+  expiryInSeconds = 25;
+  const tryLockResponse1 = await client.lock.tryLock(storeName, resourceId, lockOwner, expiryInSeconds);
+  console.log(`Acquiring lock on ${storeName}, ${resourceId} as owner: ${lockOwner}`);
```
Comment (shubham1172) [inline examples/distributedLock/TryLockApplication/src/index.ts:49]: This log should be before the actual step (applicable at other places too)

**Review by shubham1172 — CHANGES_REQUESTED** [review]

File: test/e2e/grpc/client.test.ts (hunk 14 above)
Comment (amulyavarote) [reply test/e2e/grpc/client.test.ts:359]: No it is not. The tests are complete.

File: examples/distributedLock/TryLockApplication/src/index.ts (hunk 19 above)
Comment (amulyavarote) [reply examples/distributedLock/TryLockApplication/src/index.ts:52]: It is mainly testing the different scenarios. We have to make the first process sleep for some time.

Comment (shubham1172) [reply examples/distributedLock/TryLockApplication/src/index.ts:52]: Okay, we can maybe add some logging there, i.e. going to sleep for 20 seconds, finished process, etc.

File: test/e2e/grpc/client.test.ts (hunk 14 above)
Comment (shubham1172) [reply test/e2e/grpc/client.test.ts:359]: This says that "should not be able to unlock a lock when that lock is acquired by the other owner/process", however the test only has a "tryLock" step.

Comment (amulyavarote) [reply test/e2e/grpc/client.test.ts:359]: The test name needs modification. I will change the test name.

File: test/e2e/grpc/client.test.ts [hunk 21]
```diff
@@ -318,4 +319,47 @@ describe('grpc/client', () => {
...
+    });
+
+    it('should be able to acquire a lock when a different lock is acquired by the other owner', async () => {
+      const tryLockOne = await client.lock.tryLock("redislock", "resourceId2", "owner1", 5);
+      expect(tryLockOne.success).toEqual(true);
+      const tryLockTwo = await client.lock.tryLock("redislock", "resourceId3", "owner2", 5);
+      expect(tryLockTwo.success).toEqual(true);
+    });
+
+    it('should not be able to acquire a lock when that lock is acquired by the other owner/process', async () => {
```
Comment (shubham1172) [inline test/e2e/grpc/client.test.ts:358]: I think we are missing the test case "should not be able to unlock a lock when that lock is acquired by the other owner/process".

File: test/e2e/grpc/client.test.ts [hunk 22]
```diff
@@ -318,4 +319,47 @@ describe('grpc/client', () => {
...
+      const tryLockOne = await client.lock.tryLock("redislock", "resourceId4", "owner3", 5);
+      expect(tryLockOne.success).toEqual(true);
+      const tryLockTwo = await client.lock.tryLock("redislock", "resourceId4", "owner3", 5);
```
Comment (shubham1172) [inline test/e2e/grpc/client.test.ts:361]: Should this be another owner trying to lock the same resource?

File: src/interfaces/Client/IClientLock.ts [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+
+import { TryLockResponse } from "../../types/lock/TryLockResponse";
+import { UnlockResponse } from "../../types/lock/UnlockResponse";
+
+/**
+ * Interface to acquire lock and releasing lock
+ */
+export default interface IClientLock {
+  /**
+   *
```
Comment (shubham1172) [inline src/interfaces/Client/IClientLock.ts:22]: Can we add more context, similar to https://github.com/dapr/dapr/issues/3549#issue-970288624?
Example,
```
// Required. The lock store name,e.g. `redis`.
  string store_name = 1;

  // Required. resource_id is the lock key. e.g. `order_id_111`
  // It stands for "which resource I want to protect"
  string resource_id = 2;

  // Required. lock_owner indicate the identifier of lock owner.
  // You can generate a uuid as lock_owner.For example,in **JavaScript**:
  //
  // req.LockOwner = uuid.New().String() // js example instead
  //
  // This field is per request, not per process, so it is different for each request,
  // which aims to prevent multi-thread in the same process trying the same lock concurrently.
```

**Review by shubham1172 — COMMENTED** [review]
Please fix the build and the remaining minor comments, LGTM overall! Thank you for this PR!
